module/modmanager/manager.go

```diff
@@ -143,30 +145,84 @@ func (mgr *Manager) Handles() map[string]modlib.HandlerMap {
 	return res
 }
 
-// Add adds and starts a new resource module.
-func (mgr *Manager) Add(ctx context.Context, conf config.Module) error {
+func (mgr *Manager) lockModuleMutex(modName string) func() {
 	mgr.mu.Lock()
 	defer mgr.mu.Unlock()
-	return mgr.add(ctx, conf)
+
+	value, _ := mgr.perModMu.LoadOrStore(modName, &sync.Mutex{})
+	mu := value.(*sync.Mutex)
+	mu.Lock()
+	return mu.Unlock
 }
 
-func (mgr *Manager) add(ctx context.Context, conf config.Module) error {
+// Add adds and starts a new resource module.
+func (mgr *Manager) Add(ctx context.Context, confs ...config.Module) error {
 	if mgr.untrustedEnv {
 		return errModularResourcesDisabled
 	}
 
-	_, exists := mgr.modules[conf.Name]
-	if exists {
-		return nil
+	var (
+		wg      sync.WaitGroup
+		errs    = make([]error, len(confs))
+		newMods = make([]*module, len(confs))
+	)
+
+	func() {
+		for i, conf := range confs {
+			// this is done in config validation but partial start rules require us to check again
+			if err := conf.Validate(""); err != nil {
+				mgr.logger.CErrorw(ctx, "module config validation error; skipping", "module", conf.Name, "error", err)
+				errs[i] = err
+				return
+			}
+
+			mgr.mu.RLock()
+			_, exists := mgr.modules[conf.Name]
+			mgr.mu.RUnlock()
+			if exists {
+				// module exists already!
+				return
+			}
+
+			// setup valid, new modules in parallel
+			wg.Add(1)
+			go func(i int, conf config.Module) {
+				defer wg.Done()
+
+				mod, err := mgr.add(ctx, conf)
+				if err != nil {
+					mgr.logger.CErrorw(ctx, "error adding module", "module", conf.Name, "error", err)
+					errs[i] = err
+					return
+				}
+				newMods[i] = mod
+			}(i, conf)
+		}
+		wg.Wait()
+	}()
+
+	// register new modules
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	for _, mod := range newMods {
+		if mod != nil {
+			mgr.register(mod)
+		}
 	}
 
+	return errors.Join(errs...)
+}
+
+func (mgr *Manager) add(ctx context.Context, conf config.Module) (*module, error) {
+	defer mgr.lockModuleMutex(conf.Name)()
```

Comment:
[q] What exactly is this per-module-lock guarding against?

Reply:
it's intended to protect against concurrent `Add` operations for modules with the same name (since most the artifacts we create are based on name), but allow concurrent `Add` operations for modules with different names. 

that said, I think I might have to add this to other per-module operations as well.